source/connection_manager.c

```diff
@@ -460,16 +487,127 @@ static MQTTStatus_t prvMQTTConnect( bool xCleanSession )
     LogInfo( ( "Session present: %d\n", xSessionPresent ) );
 
     /* Resume a session if desired. */
-    if( ( xResult == MQTTSuccess ) && !xCleanSession )
+    if( ( xResult == MQTTSuccess ) && ( xCleanSession == false ) )
     {
         xResult = MQTTAgent_ResumeSession( &xGlobalMqttAgentContext, xSessionPresent );
+
+        /* Resubscribe to all the subscribed topics. */
+        if( ( xResult == MQTTSuccess ) && ( xSessionPresent == false ) )
+        {
+            xResult = prvHandleResubscribe();
+        }
     }
 
     return xResult;
 }
 
 /*-----------------------------------------------------------*/
 
+static MQTTStatus_t prvHandleResubscribe( void )
+{
+    MQTTStatus_t xResult = MQTTBadParameter;
+    int32_t lIndex = 0;
+    uint16_t usNumSubscriptions = 0U;
+
+    /* These variables need to stay in scope until command completes.
+     * Dynamically allocating memory for these will help to optimize the memory
+     * by freeing up the memory once the command is complete. In this demo, only
+     * static allocations are used. */
+    static MQTTAgentSubscribeArgs_t xSubArgs = { 0 };
+    static MQTTSubscribeInfo_t xSubInfo[ SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS ] = { 0 };
+    static CommandInfo_t xCommandParams = { 0 };
+
+    /* Loop through each subscription in the subscription list and add a subscribe
+     * command to the command queue. */
+    for( lIndex = 0; lIndex < SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS; lIndex++ )
+    {
+        /* Check if there is a subscription in the subscription list. This demo
+         * doesn't subscribe to the same topics from different tasks. If there
```

Comment:
The 2nd and 3rd sentences are confusing.  I would simply say that the loop doesn't check for duplicates.